g2o/core/base_constant_edge.hpp

```diff
@@ -0,0 +1,271 @@
+            hessianTransposed.noalias() += B.transpose() * weightedOmega * A;
+          else
+            hessian.noalias() += A.transpose() * weightedOmega * B;
+        }
+      }
+      if (toNotFixed) {
+        QuadraticFormLock lck(*to);
+
+        to->b().noalias() += B.transpose() * omega_r;
+        to->A().noalias() += B.transpose() * weightedOmega * B;
+      }
+    }
+  }
+};
+
+template <int D, typename E, typename... VertexTypes>
+void BaseConstantEdge<D, E, VertexTypes...>::linearizeOplus(JacobianWorkspace& jacobianWorkspace)
+{
+  linearizeOplus_allocate(jacobianWorkspace, std::make_integer_sequence<int, _nr_of_vertices>());
+  linearizeOplus();
+}
+
+template <int D, typename E, typename... VertexTypes>
+template<int... Ints>
+void BaseConstantEdge<D, E, VertexTypes...>::linearizeOplus_allocate(JacobianWorkspace& jacobianWorkspace, std::integer_sequence<int, Ints...>)
+{
+  ( new (&std::get<Ints>(_jacobianOplus)) JacobianType<D, VertexDimension<Ints>()>(jacobianWorkspace.workspaceForVertex(Ints), D < 0 ? _dimension : D, VertexDimension<Ints>()) , ...);
+}
+
+template <int D, typename E, typename... VertexTypes>
+template<int N>
+void BaseConstantEdge<D, E, VertexTypes...>::linearizeOplusN()
+{
+  auto& jacobianOplus = std::get<N>(_jacobianOplus);
+  auto vertex = vertexXn<N>();
+
+  const number_t delta = cst(1e-9);
+  const number_t scalar = 1 / (2*delta);
+
+  bool iNotFixed = !(vertex->fixed());
+  if (iNotFixed) {
+    QuadraticFormLock lck(*vertex);
+    // estimate the jacobian numerically
+    number_t add_vertex[VertexDimension<N>()] = {};
+
+    // add small step along the unit vector in each dimension
+    for (int d = 0; d < VertexDimension<N>(); ++d) {
+      vertex->push();
+      add_vertex[d] = delta;
+      vertex->oplus(add_vertex);
+      computeError();
+      auto errorBak = this->error();
+      vertex->pop();
+      vertex->push();
+      add_vertex[d] = -delta;
+      vertex->oplus(add_vertex);
+      computeError();
+      errorBak -= this->error();
+      vertex->pop();
+      add_vertex[d] = 0.0;
+
+      jacobianOplus.col(d) = scalar * errorBak;
+    } // end dimension
+  }
+};
+
+template <int D, typename E, typename... VertexTypes>
+template <int... Ints>
+void BaseConstantEdge<D, E, VertexTypes...>::linearizeOplusNs(std::integer_sequence<int, Ints...>)
+{
+  ( void(linearizeOplusN<Ints>()), ...);
+}
+
+template <int D, typename E, typename... VertexTypes>
+void BaseConstantEdge<D, E, VertexTypes...>::linearizeOplus()
+{
+  // todo:
+  //if (!iNotFixed && !jNotFixed)
+  //  return;
+
+  ErrorVector errorBeforeNumeric = _error;
+
+  linearizeOplusNs(std::make_integer_sequence<int, _nr_of_vertices>());
+
+  _error = errorBeforeNumeric;
+}
+
+template <int D, typename E, typename... VertexTypes>
+void BaseConstantEdge<D, E, VertexTypes...>::mapHessianMemory(number_t* d, int i, int j, bool rowMajor)
+{
+  auto f = [d](auto& hessian){ new (&hessian) std::remove_reference_t<decltype(hessian)>(d); };
```

Comment:
calling a placement new on a non-trivial type is really sketchy as it won't destroy the object that was there beforehand. You should just call a regular assignment operator/constructor.

Reply:
I copied here the logic that g2o uses in all its edge classes. I do not know exactly why this choice was made. It would be good first to check this first, and perhaps change it in a different PR, also involving the dynamic base_multi_edge.